manager/deallocator/deallocator_test.go

```diff
@@ -0,0 +1,406 @@
+		require.NoError(t, store.UpdateService(tx, service))
+	},
+		// the deallocator shouldn't do any DB updates based on this event
+		false)
+}
+
+// this tests that an update to a network that is _not_ marked it for deletion
+// doesn't do anything
+func TestNetworkNotMarkedForDeletion(t *testing.T) {
+	s := store.NewMemoryStore(nil)
+	require.NotNil(t, s)
+	defer s.Close()
+
+	network := newNetwork("network", false)
+	service := newService("service", false, network)
+	createDBObjects(t, s, network, service)
+
+	deallocator, ran := startNewDeallocator(t, s, false)
+	defer stopDeallocator(t, deallocator, ran)
+
+	updateStoreAndWaitForEvent(t, deallocator, func(tx store.Tx) {
+		network.IPAM = &api.IPAMOptions{Driver: &api.Driver{Name: "test_driver"}}
+		require.NoError(t, store.UpdateNetwork(tx, network))
+	},
+		// the deallocator shouldn't do any DB updates based on this event
+		false)
+}
+
+// this test that the deallocator also works with the "old" style of storing
+// networks directly on the  service spec (instead of the task spec)
+// TODO: as said in this package's network.go's, we should really add a helper
+// on services objects itself, and test it there instead
+func TestDeallocatorWithOldStyleNetworks(t *testing.T) {
+	s := store.NewMemoryStore(nil)
+	require.NotNil(t, s)
+	defer s.Close()
+
+	service := newService("service", true)
+
+	// add a couple of networks with the old style
+	network1 := newNetwork("network1", true)
+	network2 := newNetwork("network2", false)
+	service.Spec.Networks = append(service.Spec.Networks, newNetworkConfigs(network1, network2)...)
+	task := newTask("task", service)
+
+	createDBObjects(t, s, service, network1, network2, task)
+
+	deallocator, ran := startNewDeallocator(t, s,
+		// nothing should have been deleted
+		false)
+	defer stopDeallocator(t, deallocator, ran)
+
+	// now let's delete the one task saving it all from oblivion
+	updateStoreAndWaitForEvent(t, deallocator, func(tx store.Tx) {
+		require.NoError(t, store.DeleteTask(tx, task.ID))
+	}, true)
+
+	// the deallocator should have removed both the service and
+	// the first network, but not the second
+	s.View(func(tx store.ReadTx) {
+		require.Nil(t, store.GetService(tx, service.ID))
+		require.Nil(t, store.GetNetwork(tx, network1.ID))
+		require.NotNil(t, store.GetNetwork(tx, network2.ID))
+	})
+}
+
+// Helpers below
+
+// starts a new deallocator, and also creates a channel to retrieve the return
+// value, so that we can check later than there was no error
+func startNewDeallocator(t *testing.T, s *store.MemoryStore, expectedUpdates bool) (deallocator *Deallocator, ran chan error) {
+	deallocator = New(s)
+	deallocator.eventChan = make(chan bool)
+	ran = make(chan error)
+
+	go func() {
+		returnValue := deallocator.Run(context.Background())
+		// allows checking that `Run` does return after we've stopped
+		ran <- returnValue
+		close(ran)
+	}()
+	waitForDeallocatorEvent(t, deallocator, expectedUpdates)
```

Comment:
i was wondering how you fixed the bug here. i believe you issue an event on `notifyEventChan` after the setup phase? 

a better pattern than doing that would be to add a variable to the deallocator:

```go
type Deallocator struct {
    // other fields elided
    started chan struct{}
}
```

Then, after the deallocator has started, you do:

```go
close(started)
```

Finally, instead of passing `expectedUpdates`, you should just check if the db state is in the state you want.

Reply:
Honestly not sold on this one. Adds yet another field for no great reason, and checking whether the deallocator actually performed any DB operations is an additional check on top of just checking that the DB is in the state we expect (which we also do).